proxy.go

```diff
@@ -0,0 +1,32 @@
+package proxy
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/getlantern/golog"
+	"github.com/getlantern/ops"
+)
+
+var (
+	log = golog.LoggerFor("proxy")
+)
+
+type DialFunc func(network, addr string) (conn net.Conn, err error)
+
+type Handler func(op ops.Op, w http.ResponseWriter, req *http.Request)
+
+func isUnexpected(err error) bool {
+	text := err.Error()
+	return !strings.HasSuffix(text, "EOF") && !strings.Contains(text, "use of closed network connection") && !strings.Contains(text, "Use of idled network connection")
+}
+
+func addIdleKeepAlive(header http.Header, idleTimeout time.Duration) {
+	if idleTimeout > 0 {
+		// Tell the client when we're going to time out due to idle connections
```

Comment:
The client can learn about this the hard way, but we're trying to be nice. Including the header hopefully avoids the scenario where the client tries to send a request right right at the end of the idle timeout.

We use this library in both the Lantern client and in http-proxy, so the client here is both the browser and Lantern itself, in which case it's really http.Transport.